odbc_fdw.c

```diff
@@ -1367,3 +1578,288 @@ odbcReScanForeignScan(ForeignScanState *node)
 		elog(DEBUG1, "odbcReScanForeignScan");
 	#endif
 }
+
+
+static void
+appendOption(StringInfo str, int first, const char* option_name, const char* option_value)
+{
+	if (!first)
+	{
+		appendStringInfo(str, ",\n");
+	}
+	appendStringInfo(str, "%s '%s'", option_name, option_value);
+}
+
+List *
+odbcImportForeignSchema(ImportForeignSchemaStmt *stmt, Oid serverOid)
+{
+	/* TODO: review memory management in this function; any leaks? */
+	odbcFdwOptions options;
+
+	List* create_statements = NIL;
+	List* tables = NIL;
+	List* table_columns = NIL;
+	ListCell *tables_cell;
+	ListCell *table_columns_cell;
+	RangeVar *table_rangevar;
+
+	SQLHENV env;
+	SQLHDBC dbc;
+	SQLHSTMT query_stmt;
+	SQLHSTMT columns_stmt;
+	SQLHSTMT tables_stmt;
+	SQLRETURN ret;
+	SQLSMALLINT result_columns;
+	StringInfoData col_str;
+	SQLCHAR *ColumnName;
+	SQLCHAR *TableName;
+	SQLSMALLINT NameLength;
+	SQLSMALLINT DataType;
+	SQLULEN     ColumnSize;
+	SQLSMALLINT DecimalDigits;
+	SQLSMALLINT Nullable;
+	int i;
+	StringInfoData sql_type;
+	SQLLEN indicator;
+	const char* schema_name;
+
+	#ifdef DEBUG
+		elog(DEBUG1, "odbcImportForeignSchema");
+	#endif
+
+	odbcGetOptions(serverOid, stmt->options, &options);
+
+	schema_name = get_schema_name(&options);
+
+	if (options.sql_query)
+	{
+		/* Generate foreign table for a query */
+		if (!options.table)
+		{
+			/* TODO: error */
+			elog(ERROR, "Must provide 'table' option to name the foreign table");
+		}
+
+		odbc_connection(&options, &env, &dbc);
+
+		/* Allocate a statement handle */
+		SQLAllocHandle(SQL_HANDLE_STMT, dbc, &query_stmt);
+
+		/* Retrieve a list of rows */
+		ret = SQLExecDirect(query_stmt, (SQLCHAR *) options.sql_query, SQL_NTS);
+		check_return(ret, "Executing ODBC query", query_stmt, SQL_HANDLE_STMT);
+
+		SQLNumResultCols(query_stmt, &result_columns);
+
+		initStringInfo(&col_str);
+		ColumnName = (SQLCHAR *) palloc(sizeof(SQLCHAR) * 255);
```

Comment:
of course it depends on whether we're talking about the local or the foreign name.

If we set this as a limitation, maybe better to document it. And have it in a "constant".